src/pydicom/pixels/utils.py

```diff
@@ -2098,3 +2170,246 @@ def unpack_bits(src: bytes, as_array: bool = True) -> "np.ndarray | bytes":
+        Whether to zero-pad the resulting bytes to a multiple of 2 bytes.
+
+    Returns
+    -------
+    bytes | bytearray
+        Bit-packed representation of the single frame requested.
+
+    Note
+    ----
+    In certain situations, this function will run faster and use less memory if
+    NumPy is installed. Otherwise, it will fall back to a slower method using
+    native Python functions.
+    """
+    # Pixel index of first pixel in requested frame
+    pixel_start = index * frame_length
+
+    # Pixel index one beyond the last pixel in the requested frame
+    pixel_end = pixel_start + frame_length
+
+    # Index of first byte containing pixels from the requested frame
+    byte_start = math.floor(pixel_start / 8)
+
+    # Index of the byte one beyond that containing last pixel in the requested
+    # frame
+    byte_end = math.floor((pixel_end - 1) / 8) + 1
+
+    # Mask for the final byte of the frame
+    final_byte_mask = [
+        b"\xff",  # should never be used
+        b"\x01",
+        b"\x03",
+        b"\x07",
+        b"\x0f",
+        b"\x1f",
+        b"\x3f",
+        b"\x7f",
+    ][frame_length % 8]
+
+    if pixel_start % 8 == 0:
+        # The initial frame boundary is aligned with a byte boundary, so no
+        # need to shift bits between bytes
+
+        if frame_length % 8 == 0:
+            # The end is also byte aligned so we can simply take a slice
+            out = src[byte_start:byte_end]
+
+        else:
+            # Final byte contains pixels from the next frame that need to be
+            # masked
+            out = b"".join(
+                [
+                    src[byte_start : byte_end - 1],
+                    (
+                        src[byte_end - 1] & int.from_bytes(final_byte_mask, "big")
+                    ).to_bytes(1, "big"),
+                ]
+            )
+
+    elif HAVE_NP:
+        # Start pixel is in the middle of a byte so need to shift bits between
+        # bytes. With NumPy installed, we use an efficient implementation using
+        # bitwise operations to avoid the need to unpack all the pixels
+        bit_offset = pixel_start % 8
+
+        frame_bytes = np.frombuffer(src[byte_start:byte_end], dtype=np.uint8)
+        n_output_bytes = math.ceil(frame_length / 8)
+
+        # Start with the bytes from which the least significant bits of the output
+        # are drawn
+        shifted = np.right_shift(frame_bytes[:n_output_bytes], bit_offset)
+
+        # Array of bytes from which the most significant bits of the output are
+        # drawn
+        ms_bytes = np.left_shift(frame_bytes[1:], 8 - bit_offset)
+
+        # Combine the two sets of bytes
+        np.bitwise_or(shifted[: len(ms_bytes)], ms_bytes, out=shifted[: len(ms_bytes)])
+
+        # Mask out any bits at the end that are not in this frame
+        np.bitwise_and(
+            shifted[-1:],
+            np.frombuffer(final_byte_mask, np.uint8),
+            out=shifted[-1:],
+        )
```

Comment:
Simpler to just do `shifted[-1] &= mask`?

Reply:
Addressed in [e9ea3de](https://github.com/pydicom/pydicom/pull/2216/commits/e9ea3deb538f6786f5a80e0e826d7aaf7a40678c)